node/network/collator-protocol/src/validator_side.rs

```diff
@@ -0,0 +1,1369 @@
+		}
+	}
+
+	Ok(())
+}
+
+/// Some other subsystem has reported a collator as a good one, bump reputation.
+async fn note_good_collation<Context>(
+	ctx: &mut Context,
+	state: &mut State,
+	id: CollatorId,
+) -> Result<()>
+where
+	Context: SubsystemContext<Message = CollatorProtocolMessage>
+{
+	for (peer_id, collator_id) in state.known_collators.iter() {
+		if id == *collator_id {
+			modify_reputation(ctx, peer_id.clone(), BENEFIT_NOTIFY_GOOD).await?;
+		}
+	}
+
+	Ok(())
+}
+
+/// A peer's view has changed. A number of things should be done:
+///  - Ongoing collation requests have to be cancelled.
+///  - Advertisments by this peer that are no longer relevant have to be removed.
+async fn handle_peer_view_change(
+	state: &mut State,
+	peer_id: PeerId,
+	view: View,
+) -> Result<()> {
+	let current = state.peer_views.entry(peer_id.clone()).or_default();
+
+	let removed: Vec<_> = current.difference(&view).cloned().collect();
+
+	*current = view;
+
+	if let Some(advertisments) = state.advertisments.get_mut(&peer_id) {
+		advertisments.retain(|(_, relay_parent)| !removed.contains(relay_parent));
+	}
+
+	let mut requests_to_cancel = Vec::new();
+
+	for removed in removed.into_iter() {
+		state.requested_collations.retain(|k, v| {
+			if k.0 == removed {
+				requests_to_cancel.push(*v);
+				false
+			} else {
+				true
+			}
+		});
+	}
+
+	for r in requests_to_cancel.into_iter() {
+		if let Some(per_request) = state.requests_info.remove(&r) {
+			per_request.received.send(()).map_err(|_| oneshot::Canceled)?;
+		}
+	}
+
+	Ok(())
+}
+
+/// We have received a collation.
+///  - Cancel all ongoing requests
+///  - Reply to interested parties if any
+///  - Store collation.
+async fn received_collation<Context>(
+	ctx: &mut Context,
+	state: &mut State,
+	origin: PeerId,
+	request_id: RequestId,
+	receipt: CandidateReceipt,
+	pov: PoV,
+) -> Result<()>
+where
+	Context: SubsystemContext<Message = CollatorProtocolMessage>
+{
+	let relay_parent = receipt.descriptor.relay_parent;
+	let para_id = receipt.descriptor.para_id;
+
+	if let Some(id) = state.requested_collations.remove(
+		&(relay_parent, para_id, origin.clone())
+	) {
+		if id == request_id {
+			if let Some(mut per_request) = state.requests_info.remove(&id) {
+				let _ = per_request.received.send(());
+				if let Some(collator_id) = state.known_collators.get(&origin) {
+					if let Some(ref mut result) = per_request.result {
+						let _ = result.send((collator_id.clone(), receipt.clone(), pov.clone())).await;
+					}
+
+					state.collations
+						.entry((relay_parent, para_id))
+						.or_default()
+						.push((collator_id.clone(), receipt, pov));
+				}
+			}
+		}
+	} else {
+		// TODO: This is tricky. If our chain has moved on, we have already canceled
+		// the relevant request and removed it from the map; so and we are not expecting
+		// this reply although technically it is not a malicious behaviur.
+		modify_reputation(ctx, origin, COST_UNEXPECTED_MESSAGE).await?;
```

Comment:
The current behavior seems plausible to me. Even so, should remove the TODO or link an issue for future discussion of what the behavior should be.

Reply:
yep, mostly TODOs exist in the PR at places I have no idea what to do about, so

Reply:
One possible behavior would be to keep around requests for one block and then drop them. If a request arrives from an old block we know it is expired but don't need to report as the response is genuine.

However I can see the current jobs system making that hard. @coriolinus This is the kind of stuff I was referring to when I mentioned that the jobs system is kind of fragile. We should have a notion of transitions (parent -> child) and global data in the subsystem library.

Reply:
@rphmeier agree that it would be handy to have parent relation information in the active set update broadcasts, so it's possible to do things like "keep information for N blocks". However, I think that's a feature of its own which should probably be moved to its own issue so as not to delay this PR.